sql/analyzer/optimization_rules.go

```diff
@@ -355,3 +355,118 @@ func isTrue(e sql.Expression) bool {
 	}
 	return false
 }
+
+// pushNotFilters applies De'Morgan's laws to push NOT expressions as low
+// in expression trees as possible and inverts NOT leaf expressions.
+// ref: https://en.wikipedia.org/wiki/De_Morgan%27s_laws
+// note: the output tree identity will not be accurate
+func pushNotFilters(_ *sql.Context, _ *Analyzer, n sql.Node, _ *plan.Scope, _ RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	return transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		var e sql.Expression
+		var err error
+		switch n := n.(type) {
+		case *plan.Filter:
+			e, err = pushNotFiltersHelper(n.Expression)
+		case *plan.JoinNode:
+			if n.Filter != nil {
+				e, err = pushNotFiltersHelper(n.Filter)
+			}
+		default:
+			return n, transform.SameTree, nil
+		}
+		if err != nil {
+			return n, transform.SameTree, nil
+		}
+		ret, err := n.(sql.Expressioner).WithExpressions(e)
+		if err != nil {
+			return n, transform.SameTree, nil
+		}
+		return ret, transform.NewTree, nil
+	})
+}
+
+func pushNotFiltersHelper(e sql.Expression) (sql.Expression, error) {
+	// NOT(NOT(c))=>c
+	if not, _ := e.(*expression.Not); not != nil {
+		if f, _ := not.Child.(*expression.Not); f != nil {
+			return pushNotFiltersHelper(f.Child)
+		}
+	}
+
+	// NOT(AND(left,right))=>OR(NOT(left), NOT(right))
+	if not, _ := e.(*expression.Not); not != nil {
+		if f, _ := not.Child.(*expression.And); f != nil {
+			return pushNotFiltersHelper(expression.NewOr(expression.NewNot(f.Left), expression.NewNot(f.Right)))
+		}
+	}
+
+	// NOT(OR(left,right))=>AND(NOT(left), NOT(right))
+	if not, _ := e.(*expression.Not); not != nil {
+		if f, _ := not.Child.(*expression.Or); f != nil {
+			return pushNotFiltersHelper(expression.NewAnd(expression.NewNot(f.Left), expression.NewNot(f.Right)))
+		}
+	}
+
+	// TODO this doesn't appear to be correct
+	// NOT(ISNULL(c))=>OR(LT(c,NULL), GT(c,NULL))
```

Comment:
just out of curiosity... does splitting `NOT(ISNULL(c))` into an `OR` expression help with plan optimization? Seems like `ISNULL` is simpler than two comparisons, but I figured you must have had something in mind here. I guess we could potentially short circuit if the first expression in the `OR` was true, but that still seems comparable to comparing to NULL. 

It seems like this on breaks down because of NULL comparison. Anything compared with NULL will be NULL, so it would always reduce to `NULL OR NULL` and end up being neither true or false.

Reply:
Yeah the split seems to break correctness also. Matching the NOT->IS_NULL pattern on my other branch wasn't such a big deal.